core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -243,27 +245,42 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
     // Launch tasks returned by a set of resource offers
     private def launchTasks(tasks: Seq[Seq[TaskDescription]]) {
-      for (task <- tasks.flatten) {
-        val serializedTask = TaskDescription.encode(task)
-        if (serializedTask.limit >= maxRpcMessageSize) {
-          scheduler.taskIdToTaskSetManager.get(task.taskId).foreach { taskSetMgr =>
-            try {
-              var msg = "Serialized task %s:%d was %d bytes, which exceeds max allowed: " +
-                "spark.rpc.message.maxSize (%d bytes). Consider increasing " +
-                "spark.rpc.message.maxSize or using broadcast variables for large values."
-              msg = msg.format(task.taskId, task.index, serializedTask.limit, maxRpcMessageSize)
-              taskSetMgr.abort(msg)
-            } catch {
-              case e: Exception => logError("Exception in error callback", e)
-            }
+      val serializedTasks = tasks.flatten.map { task =>
+        var serializedTask: ByteBuffer = null
+        try {
+          serializedTask = TaskDescription.encode(task, task.serializedTask)
+          if (serializedTask.limit >= maxRpcMessageSize) {
+            val msg = "Serialized task %s:%d was %d bytes, which exceeds max allowed: " +
+              "spark.rpc.message.maxSize (%d bytes). Consider increasing " +
+              "spark.rpc.message.maxSize or using broadcast variables for large values."
+            abortTaskSetManager(scheduler, task.taskId,
+              msg.format(task.taskId, task.index, serializedTask.limit, maxRpcMessageSize))
+            serializedTask = null
+          } else if (serializedTask.limit > TaskSetManager.TASK_SIZE_TO_WARN_KB * 1024) {
+            scheduler.taskIdToTaskSetManager.get(task.taskId).filterNot(_.emittedTaskSizeWarning).
+              foreach { taskSetMgr =>
+                taskSetMgr.emittedTaskSizeWarning = true
+                val stageId = taskSetMgr.taskSet.stageId
+                logWarning(s"Stage $stageId contains a task of very large size " +
+                  s"(${serializedTask.limit / 1024} KB). The maximum recommended task size is " +
+                  s"${TaskSetManager.TASK_SIZE_TO_WARN_KB} KB.")
+              }
           }
+        } catch {
+          case NonFatal(e) =>
+            abortTaskSetManager(scheduler, task.taskId,
+              s"Failed to serialize task ${task.taskId}, not attempting to retry it.", Some(e))
         }
-        else {
+        (task, serializedTask)
+      }
+
+      if (!serializedTasks.exists(b => b._2 eq null)) {
```

Comment:
Can you start by writing a unit test for the issue Imran brought up?  That will make it easier to ensure that the issue is fixed.